### PR TITLE
[TT-1218] result table & workflow parser tools

### PR DESCRIPTION
**Pull request by Tofel**
Here we add two new tools:
* `asciitable` creates a 2 column ASCII table based on provided json. It's capable of appending new data to existing table and adjusting width of existing rows to new max width. Second table supports only two values: ✅ and ❌. The main use of this tool is to present results in an aggregated way (for example in job summary). Example:
```
+------------------------------+--------+
| EVM Implementation           | Result |
+------------------------------+--------+
|                 vrfv2                 |
+------------------------------+--------+
| ethereum/client-go:v1.14.5   | √      |
| ethereum/client-go:v1.14.4   | √      |
| ethereum/client-go:v1.14.3   | √      |
| thorax/erigon:v2.59.3        | √      |
| thorax/erigon:v2.59.2        | √      |
| thorax/erigon:v2.59.1        | √      |
+------------------------------+--------+
```
* `workflowresultparser` is a small tool that grabs results of jobs described by a regex pattern from a given workflow. It supports pagination. It's main purpose is fetching of data that's later used to create above mentioned ASCII table and input for slack notifications. Currently that data fetching is done in bash and doesn't support pagination.

Both tools have some flexibility, but not too much as they were tailored to rather specific use cases.

---
**Below is a summarization created by an LLM (gpt-4-0125-preview). Be mindful of hallucinations and verify accuracy.**

## Why
The changes introduce two new tools, `asciitable` and `workflowresultparser`, to the codebase, including their implementation, tests, documentation, and build configuration. These tools are designed to enhance the project's capabilities in generating ASCII tables from JSON data and parsing GitHub Actions workflow results, respectively.

## What
- **.github/workflows/lint.yaml**  
  - Added two new tools `asciitable` and `workflowresultparser` to the linting workflow, ensuring code quality and consistency.
- **.github/workflows/test.yaml**  
  - Included `asciitable` and `workflowresultparser` in the unit testing workflow, verifying their functionality and correctness.
- **tools/asciitable/Makefile**  
  - Created a new Makefile for the `asciitable` tool, defining `lint` and `test_unit` targets for code quality checks and unit testing.
- **tools/asciitable/README.md**  
  - Added a README file for the `asciitable` tool, providing usage instructions, examples, and a detailed description of the tool's functionality.
- **tools/asciitable/go.mod** and **tools/asciitable/go.sum**  
  - Initialized Go module for `asciitable`, specifying dependencies.
- **tools/asciitable/main.go**  
  - Implemented the main logic for the `asciitable` tool, including JSON parsing, table formatting, and result output.
- **tools/asciitable/main_test.go**  
  - Added unit tests for the `asciitable` tool, covering various scenarios and ensuring the correctness of its functionality.
- **tools/asciitable/package.json**  
  - Created a package.json file for `asciitable`, defining basic metadata such as the tool's name and version.
- **tools/workflowresultparser/Makefile**  
  - Created a new Makefile for the `workflowresultparser` tool, defining `lint` and `test_unit` targets.
- **tools/workflowresultparser/README.md**  
  - Added a README file for the `workflowresultparser` tool, detailing its purpose, usage, and command-line options.
- **tools/workflowresultparser/go.mod** and **tools/workflowresultparser/go.sum**  
  - Initialized Go module for `workflowresultparser`, listing its dependencies.
- **tools/workflowresultparser/main.go**  
  - Implemented the main functionality of the `workflowresultparser` tool, including GitHub API interaction, job filtering, and result processing.
- **tools/workflowresultparser/main_test.go**  
  - Added unit tests for the `workflowresultparser` tool, ensuring its reliability and accuracy in parsing workflow results.
- **tools/workflowresultparser/package.json**  
  - Created a package.json file for `workflowresultparser`, providing basic information such as the tool's name and version.

### CODE REVIEW

File: tools/workflowresultparser/main.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	re := regexp.MustCompile(*jobNameRegex)
+	for _, job := range jobs {
+		if re.MatchString(job.Name) {
+			for _, step := range job.Steps {
+				if step.Name == "Run Tests" {
+					conclusion := ":x:"
+					if step.Conclusion == "success" {
+						conclusion = ":white_check_mark:"
+					}
+					captureGroup := re.FindStringSubmatch(job.Name)[1]
```
Comment (skudasov) [inline tools/workflowresultparser/main.go:111]: Can we have an empty capture group here?

File: tools/workflowresultparser/main.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	return parsedResults, nil
+}
+
+func processResults(parsedResults []ParsedResult, namedKey, jobNameRegex, workflowRunID, outputFile *string) error {
+	results := ResultsMap{}
+
+	if *outputFile != "" {
+		if _, statErr := os.Stat(*outputFile); statErr == nil {
+			existingData, readErr := os.ReadFile(*outputFile)
+			if readErr == nil {
```
Comment (skudasov) [inline tools/workflowresultparser/main.go:135]: Maybe just do if err != nil first? Easier to read

File: tools/workflowresultparser/main.go (hunk 1 above)
Comment (Tofel) [reply tools/workflowresultparser/main.go:111]: I was also worried about it, but I have a test that doesn't match anything and nothing panics, which is a bit surprising, tbh

I was thinking also about adding a input job regex validation to make sure it contains at least 1 capture group

File: tools/asciitable/main_test.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+import (
+	"flag"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestReadJSONFile(t *testing.T) {
```
Comment (skudasov) [inline tools/asciitable/main_test.go:11]: We can remove that, we are just testing the stdlib here

File: tools/asciitable/main.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+			if len(job.Cap) > maxFirstColumnLen {
+				maxFirstColumnLen = len(job.Cap)
+			}
+		}
+	}
+
+	return maxFirstColumnLen, maxSecondColumnLen
+}
+
+func writeResultsToFile(fileName string, firstColumnHeader, secondColumnHeader, currentSection string, jobs []ParsedResult) error {
```
Comment (skudasov) [inline tools/asciitable/main.go:61]: Can we use https://github.com/olekukonko/tablewriter instead? Okay if we need it urgently but seems like we can replace this code with a more solid library

Comment (Tofel) [reply tools/asciitable/main.go:57]: will create a low prio ticket, wdyt?

**Review by skudasov — APPROVED** [review]

**Pull request merged**